### ISSUE TITLE
Symbol search for symbol in dependency

### ISSUE THREAD

**Issue by obsgolem** [rust-lang/rust-analyzer#13938]
I frequently find myself wanting to explore the source of a dependency in order to understand how to use it. Go to definition works great for this if I have code that references the thing I want to explore, but sometimes I don't. For this usecase, I want to use go to symbol instead. A viable alternative is to search for the name of the symbol using autocomplete, then use go to definition from there, but this is an extra step.

**Comment by flodiebold** [comment]
The workspace symbol search supports this, see [the manual](https://rust-analyzer.github.io/manual.html#workspace-symbol). In non-VSCode editors, you can just type `Foo*` to search for `Foo` in dependencies; sadly VSCode removes the star, so this doesn't work, so currently you have to change the configuration to search in dependencies.

**Comment by Veykril** [comment]
(Sorry for hijacking but) regarding VSCode I wonder if we can just overwrite the whole symbol search infra, given we can inject middlewares, to get around VSCode stripping symbols for no reason 🤔 That is we sent the corresponding request ourselves instead.

**Comment by flodiebold** [comment]
Did we ever report/ask the VSCode team about the symbol stripping, by the way?

**Comment by Veykril** [comment]
I don't think so, also my `for no reason` was ill-mannered. I do recall them having a reason for this (that is reserving some symbols in there iirc, I saw a comment regarding that at some point).

**Comment by ChayimFriedman2** [comment]
@Veykril Apparently no! VSCode only strips `*` and `#` (the conditions aren't entirely clear to me though, but it seems to not strip at least `:` in the middle of the query, which is important for me because I want to make it better with this).

**Comment by Veykril** [comment]
Huh I could swear I checked `:` specifically back when I was investigating this (as the application for that is obvious ;)

**Comment by Veykril** [comment]
Ah I was misremembering the exact behavior, we do get the full query, but VSCode does some weird filtering on the results wrt to the query string https://github.com/microsoft/vscode/issues/171320

**Comment by ChayimFriedman2** [comment]
So **after** implementing searching for `foo::bar` I discovered VSCode *does* filter the results for some unknown reason. This is so annoying that I think I'm going to contribute to VSCode.

**Comment by ChayimFriedman2** [comment]
Update: I implemented a fix in VSCode, but it has some bad scoring - e.g. when I search for `string::string*`, first it shows `StringTooLongError` from `pulldown-cmark`, and the obvious `std::string::String` only comes 7th. A proper fix for that will require changes to the LSP, so this is what I'm going to focus on now.

**Comment by Veykril** [comment]
There is value in landing this either way, at least for other editors that don't do weird scoring like vscode

**Comment by ChayimFriedman2** [comment]
I have it implemented (without tests though, I tested it manually), but I don't like the idea of shipping something that the majority of our users (including me) cannot use. I prefer to wait to see what happens with my LSP suggestion (why BTW is [here](https://github.com/microsoft/language-server-protocol/issues/2091)).
